.github/workflows/test.yml

```diff
@@ -3,41 +3,137 @@ name: CI
 on: [push, pull_request]
 
 jobs:
-    unix-test:
+    build-test:
         runs-on: ${{ matrix.os }}
         strategy:
             fail-fast: false
             matrix:
                 os: [ubuntu-latest, macos-latest]
+                build: [general, webrender]
 
         steps:
             - uses: actions/checkout@v2
 
+            - name: rust setup
+              id: rust-setup
+              run: |
+                  rustup install $(cat rust-toolchain)
+                  rustup component add rustfmt-preview
+                  echo "::set-output name=week::$(/bin/date -u '+%Yy_%mm_%Ww')"
+                  echo "::set-output name=date::$(/bin/date)"
+
+            - name: rust cache
+              uses: actions/cache@v2
+              with:
+                  path: |
+                      ~/.cargo/registry/cache
+                      ~/.cargo/git
+                      ./rust_src/target
+                  key: ${{ runner.os }}-${{ matrix.build }}-rust-${{ hashFiles('**/rust-toolchain', '**/Cargo.toml*', '**/rustfmt.toml') }}-${{ steps.rust-setup.outputs.week }}
+                  restore-keys: |
```

Comment:
Do we need the `restore-keys`? It seems that the cache is only stored using the key `${{ runner.os }}-${{ matrix.build }}-rust-${{ hashFiles('**/rust-toolchain', '**/Cargo.toml*', '**/rustfmt.toml') }}-${{ steps.rust-setup.outputs.week }}` that would not match the second/third entries.

Reply:
As previously mentioned, this is a method for automatically updating the cache.
Even if the toolchain changes, we don't have to download it all packages from the beginning.